gqcp/include/QCModel/HF/GHF.hpp

```diff
@@ -0,0 +1,395 @@
+    }
+
+
+    /**
+     *  @param M            the number of spinors
+     *  @param N            the number of electrons
+     *
+     *  @return the GHF 1-DM expressed in an orthonormal spinor basis
+     */
+    static OneDM<Scalar> calculateOrthonormalBasis1DM(const size_t M, const size_t N) {
+
+        // The 1-DM for GHF looks like (for M=5, N=3)
+        //    1  0  0  0  0
+        //    0  1  0  0  0
+        //    0  0  1  0  0
+        //    0  0  0  0  0
+        //    0  0  0  0  0
+
+        OneDM<double> D_MO = OneDM<double>::Zero(M, M);
+        D_MO.topLeftCorner(N, N) = SquareMatrix<double>::Identity(N, N);
+
+        return D_MO;
+    }
+
+
+    /**
+     *  @param M            the number of spinors
+     *  @param N            the number of electrons
+     * 
+     *  @return the implicit (i.e. with ascending and contiguous orbital indices) occupied-virtual orbital space that corresponds to these GHF model parameters
+     */
+    static OrbitalSpace orbitalSpace(const size_t M, const size_t N) {
+
+        return OrbitalSpace::Implicit({{OccupationType::k_occupied, N}, {OccupationType::k_virtual, M - N}});
+    }
+
+
+    /*
+     *  PUBLIC METHODS
+     */
+
+    /**
+     *  @param spin_op                      the electronic spin operator
+     *  @param S                            the (spin-blocked) overlap matrix of the underlying AO bases
+     * 
+     *  @return the expectation value of the electronic spin operator
+     */
+    Vector<complex, 3> calculateExpectationValueOf(const ElectronicSpinOperator& spin_op, const QCMatrix<Scalar>& S) const {
+
+        // Prepare some variables.
+        const auto M = this->numberOfSpinors();
+        const MatrixX<complex> C_alpha = this->coefficientMatrix().topRows(M / 2);
+        const MatrixX<complex> C_beta = this->coefficientMatrix().bottomRows(M / 2);
+        const QCMatrix<complex> S_AO = S.topLeftCorner(M / 2, M / 2);  // assume equal for alpha and beta
+
+        // Calculate overlaps between the alpha- and beta-spinors.
+        const MatrixX<complex> overlap_aa = C_alpha.adjoint() * S_AO * C_alpha;
+        const MatrixX<complex> overlap_ab = C_alpha.adjoint() * S_AO * C_beta;
+        const MatrixX<complex> overlap_ba = overlap_ab.adjoint();
+        const MatrixX<complex> overlap_bb = C_beta.adjoint() * S_AO * C_beta;
+
+
+        // A KISS implementation of the expectation value of S, from knowdes.
+        complex s_x {0.0};
+        for (size_t I = 0; I < this->numberOfElectrons(); I++) {  // loop over occupied spinors
+            s_x += overlap_ab(I, I).real();
+        }
+
+        complex s_y {0.0};
+        for (size_t I = 0; I < this->numberOfElectrons(); I++) {  // loop over occupied spinors
+            s_y += overlap_ba(I, I).imag();
+        }
+
+        complex s_z {0.0};
+        for (size_t I = 0; I < this->numberOfElectrons(); I++) {  // loop over occupied spinors
+            s_z += 0.5 * (overlap_aa(I, I) - overlap_bb(I, I));
+        }
```

Comment:
I think it's better to create an `OrbitalSpace` for the occupied spinors and loop over `orbitalSpace.indices(OccupationType::Occupied)`.

Reply:
Orbital space consists of vectors, so by doing this I would have to add a lot of code. Creating vectors containing the occupied & virtual indices, creating the orbital space itself & finally I would have to find a way to loop over the elements of the `occupied indices vector`. (An iterator?) Don't you think this would make things less clear?